cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+                                          Function<Vertx, CruiseControlApi> cruiseControlClientProvider, String ccHost) {
+        super(vertx, KafkaRebalance.RESOURCE_KIND, supplier.kafkaRebalanceOperator, supplier.metricsProvider);
+        this.pfa = pfa;
+        this.kafkaRebalanceOperator = supplier.kafkaRebalanceOperator;
+        this.kafkaOperator = supplier.kafkaOperator;
+        this.cruiseControlClientProvider = cruiseControlClientProvider;
+        this.ccHost = ccHost;
+    }
+
+    /**
+     * Create a watch on {@code KafkaRebalance} in the given {@code watchNamespaceOrWildcard}.
+     *
+     * @param watchNamespaceOrWildcard The namespace to watch, or "*" to watch all namespaces.
+     * @return A future which completes when the watch has been set up.
+     */
+    public Future<Void> createRebalanceWatch(String watchNamespaceOrWildcard) {
+
+        return Util.async(this.vertx, () -> {
+            this.kafkaRebalanceOperator.watch(watchNamespaceOrWildcard, new Watcher<KafkaRebalance>() {
+                @Override
+                public void eventReceived(Action action, KafkaRebalance kafkaRebalance) {
+                    Reconciliation reconciliation = new Reconciliation("kafkarebalance-watch", kafkaRebalance.getKind(),
+                            kafkaRebalance.getMetadata().getNamespace(), kafkaRebalance.getMetadata().getName());
+
+                    log.debug("{}: EventReceived {} on {} with status [{}] and {}={}", reconciliation, action,
+                            kafkaRebalance.getMetadata().getName(),
+                            kafkaRebalance.getStatus() != null ? kafkaRebalance.getStatus().getConditions().get(0).getType() : null,
+                            ANNO_STRIMZI_IO_REBALANCE, rawRebalanceAnnotation(kafkaRebalance));
+
+                    withLock(reconciliation, LOCK_TIMEOUT_MS,
+                        () -> reconcileRebalance(reconciliation, action == Action.DELETED ? null : kafkaRebalance));
+                }
+
+                @Override
+                public void onClose(KubernetesClientException e) {
+                    if (e != null) {
+                        throw e;
+                    }
+                }
+
+            });
+            return null;
+        });
+    }
+
+    private static Future<KafkaRebalance> updateStatus(KafkaRebalance clusterRebalance,
+                                                      KafkaRebalanceStatus desiredStatus,
+                                                      CrdOperator<KubernetesClient, KafkaRebalance, KafkaRebalanceList, DoneableKafkaRebalance> clusterRebalanceOperations,
+                                                      Throwable e) {
+        if (e != null) {
+            StatusUtils.setStatusConditionAndObservedGeneration(clusterRebalance, desiredStatus, e);
+        } else if (desiredStatus.getConditions() != null && desiredStatus.getConditions().get(0).getType() != null) {
```

Comment:
I still feel, like Jakub, that this is fragile. Let's do the right thing and fnd the condition that we're interested in rather than assuming that there is a nonempty list of conditions and that the one we want is the first.

Reply:
I still don't understand what you mean guys. Which condition should I search for? What should be the query and the parameter? There is only one condition, that's it. Can you elaborate please?

Reply:
Today there's a single condition with `type: Status`. If the conditions are not empty we know it must be at index 0. We're not checking for emptiness, but if our code is perfect maybe the state machine guarantees it's never empty. But maybe our code is not perfect. If we later decided to add another condition with `type: Warning` then we can't be sure so certain it'll be at index 0 any more. So let's write a method which searches the conditions for the first condition of `type: Status`, with null checks on all the getters. It's just more robust code.

Reply:
Maybe this is the misunderstanding! `type` is not `Status`, but `type` contains the actual status ... so `type: PendingProposal`, `type: ProposalReady`, `type: Rebalancing` ... so we are storing the state machine status in the `type` field of the only one condition we have there.

Reply:
I think Tom means that in future there could be more than one condition and that we cannot guarantee that the CC state condition will be at position 0 in the list returned by `getConditions()`. If there was ever more than one condtion we could end up checking the type of a different condition.

We should have a function that loops through the condition list and searches for one that has a status type matching any of those defined in the state enum and then return that type.

Reply:
What's the value of that? At this point, we should completely change the structure of the status and set `type: State` then adding a new field `state` with the actual value as `PendingProposal`, `Ready`, .... while today something like `Ready` is just set in the `type` field.

Reply:
We don't meed to do that. This issue is what if `desiredStatus.getConditions().get(0)` is for a different condition where `getType()` is null but `desiredStatus.getConditions().get(1).getType()` is our condition and is not null?

We just need something like `findCCStateCondition(desiredStatus.getConditions()).getType() != null` or something similar.

Reply:
I'm in agreement here, this may well be fragile. implementing a method like Tom described should fix this.

Reply:
I have added a function to search for a condition which has a type matching one of the enum states and return that. This should prevent issues with multiple conditions in the future.

Reply:
We are using the `type` field as the state of the state machine, what's the filter we should use to get THE condition? I still don't understand. It makes sense to me if having more conditions, one has `type: State` so filtering on it and then getting the state from a `state` field but it's not the case here.

Reply:
We addressed this in the latest commit. Now `type` is always `State` in the condition, while the `status` field contains the state machine state (i.e. PendingProposal, Rebalancing, Ready and so on).